vyper/semantics/analysis/utils.py

```diff
@@ -645,15 +645,11 @@ def check_modifiability(node: vy_ast.VyperNode, modifiability: Modifiability) ->
         return all(check_modifiability(item, modifiability) for item in node.elements)
 
     if isinstance(node, vy_ast.Call):
-        args = node.args
-        if len(args) == 1 and isinstance(args[0], vy_ast.Dict):
-            return all(check_modifiability(v, modifiability) for v in args[0].values)
-
         call_type = get_exact_type_from_node(node.func)
 
-        # builtins
-        call_type_modifiability = getattr(call_type, "_modifiability", Modifiability.MODIFIABLE)
-        return call_type_modifiability >= modifiability
+        # structs and interfaces
+        if hasattr(call_type, "check_modifiability_for_call"):
```

Comment:
as a general comment, i'd like to move these hasattr checks to be more like
```python
if call_type._is_callable:
    call_type.check_modifiability_for_call(...)
 ```
that way we impose more structure on the APIs of callable types. but i think that can be pushed to "future work"